### PR TITLE
[IPU] Allow poptorch.Options to override Trainer

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Follow-on from #8223. This will follow along from the conversation I had with @davors72, allowing the poptorch.Options that are passed into the IPUPlugin to be the source of truth, allowing the user to not have to pass any flags directly to the Trainer. I.e:

```python
inference_opts = poptorch.Options()
inference_opts.replicationFactor(4)
inference_opts.deviceIterations(2)
inference_opts.Training.gradientAccumulation(1)

training_opts = poptorch.Options()
training_opts.replicationFactor(4)
training_opts.deviceIterations(2)
training_opts.Training.gradientAccumulation(2)

trainer = pl.Trainer(max_epochs=2, plugins=IPUPlugin(training_opts=training_opts, inference_opts=inference_opts))
```

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pl_examples/ipu_examples/mnist.py [hunk 1]
```diff
@@ -79,11 +81,20 @@ def configure_optimizers(self):
...
 
 if __name__ == '__main__':
-    dm = MNISTDataModule(batch_size=32)
+    dm = MNISTDataModule(batch_size=64)
 
     model = LitClassifier()
+    inference_opts = poptorch.Options()
+    inference_opts.replicationFactor(4)
+    inference_opts.deviceIterations(2)
+    inference_opts.Training.gradientAccumulation(1)
```
Comment (davors72) [inline pl_examples/ipu_examples/mnist.py:90]: Probably shouldn't set this. It's harmless but people might think you need to do this if it's in the example

File: pytorch_lightning/plugins/training_type/ipu.py [hunk 2]
```diff
@@ -127,16 +126,24 @@ def pre_dispatch(self) -> None:
...
             self.poptorch_models[x] = model
+        self._handle_gradient_accumulation_steps()
 
     @property
-    def replication_factor(self):
+    def replication_factor(self) -> int:
+        # This should be set via the opts not the parallel devices
+        if self.lightning_module and self.lightning_module.trainer.training and self._training_opts is not None:
+            return self._training_opts.replication_factor
+        elif self._training_opts is not None:
```
Comment (davors72) [inline pytorch_lightning/plugins/training_type/ipu.py:136]: Not sure this logic is right here. Shouldn't this be: `self.lightning_module.trainer.training and self._training_opts is not None:`

File: pl_examples/ipu_examples/mnist.py (hunk 1 above)
Comment (SeanNaren) [reply pl_examples/ipu_examples/mnist.py:90]: thanks! yep will revert all the changes here, just using it to ensure manual plugin support works

File: pytorch_lightning/plugins/training_type/ipu.py (hunk 2 above)
Comment (SeanNaren) [reply pytorch_lightning/plugins/training_type/ipu.py:136]: Yeah sounds right, I made it like this as the lightning module when this is called may not be a reference yet (we haven't called fit on the model). Thus we default to the training opts. Need to make this better though

File: pytorch_lightning/plugins/training_type/ipu.py [hunk 3]
```diff
@@ -66,7 +66,7 @@ def __init__(
         device_iterations: int = 1,
         autoreport: bool = True,
         autoreport_dir: Optional[str] = None,
-        parallel_devices: Optional[List[torch.device]] = None,
+        parallel_devices: Optional[List[torch.device]] = (),
```
Comment (SeanNaren) [inline pytorch_lightning/plugins/training_type/ipu.py:69]: This is a bit hacky, I changed this to an empty tuple as this will prevent a crash since parallel devices haven't been defined when the plugin is passed and no `ipus` have also been defined in the Trainer.

File: pytorch_lightning/plugins/training_type/ipu.py [hunk 4]
```diff
@@ -129,10 +129,18 @@ def pre_dispatch(self) -> None:
         self._handle_gradient_accumulation_steps()
 
     @property
-    def replication_factor(self):
+    def replication_factor(self) -> int:
+        if not self.lightning_module:
+            # The plugin has been passed in by the user and has not been connected to the Trainer.
+            # Check if the user has passed in custom poptorch.Options to infer number of IPUs being used.
+            # In this scenario we prioritize the training options.
```
Comment (SeanNaren) [inline pytorch_lightning/plugins/training_type/ipu.py:136]: cc @davors72 for stdout logging purposes, I default to the training opts replication factor. Obviously if this is setup different for inference and training it doesn't matter as this is just for logging purposes!

File: pytorch_lightning/plugins/training_type/ipu.py [hunk 5]
```diff
@@ -147,71 +155,41 @@ def _create_opts(self, training: bool):
...
     def on_reset_test_dataloader(self, dataloader: Union[Iterable, DataLoader]) -> Union[Iterable, DataLoader]:
-        return self.process_dataloader(dataloader)
+        return self._process_dataloader(dataloader, is_training=False)
 
     def on_reset_predict_dataloader(self, dataloader: Union[Iterable, DataLoader]) -> Union[Iterable, DataLoader]:
-        return self.process_dataloader(dataloader)
+        return self._process_dataloader(dataloader, is_training=False)
 
-    def process_dataloader(self, dataloader: Union[Iterable, DataLoader]) -> Union[Iterable, DataLoader]:
+    def _process_dataloader(self, dataloader: Union[Iterable, DataLoader],
```
Comment (SeanNaren) [inline pytorch_lightning/plugins/training_type/ipu.py:182]: Process dataloader was called every time regardless of stage. By making this a protected function and calling it explicitly in the hook, I was able to get the appropriate behaviour.

File: pytorch_lightning/plugins/training_type/ipu.py [hunk 6]
```diff
@@ -147,71 +155,41 @@ def _create_opts(self, training: bool):
...
+                            is_training: bool) -> Union[Iterable, DataLoader]:
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/ipu.py:183]:
```suggestion
                            is_training: bool,) -> Union[Iterable, DataLoader]:
```
can try this, maybe the auto-formatter will put it on a new line

**Review by awaelchli — APPROVED** [review]

**Review by tchaton — APPROVED** [review]

**Review by justusschock — APPROVED** [review]
